16-encoding-categorical-data.Rmd

```diff
@@ -1,8 +1,356 @@
+  slice(1) %>% 
+  pull(.row)
+recipe(~Bldg_Type, data = ames_train) %>% 
+  step_mutate(`Raw Data` = Bldg_Type) %>% 
+  step_dummy(Bldg_Type, naming = function(var, lvl, ordinal = FALSE, sep = "_") lvl) %>% 
+  prep() %>% 
+  bake(ames_train) %>% 
+  slice(show_rows) %>% 
+  arrange(`Raw Data`) %>% 
+  kable() %>% 
+  kable_styling(full_width = FALSE)
+```
+
+Many model implementations require such a transformation to a numeric representation for categorical data. 
+
+:::rmdnote
+Appendix \@ref(pre-proc-table) presents a table of recommended preprocessing techniques for different models; notice how many of the models in the table require a numeric encoding for all predictors.
+:::
+
+However, for many realistic data sets, straightforward dummy variables are not a good fit. This often happens because there are **too many** categories or there are **new** categories at prediction time. In this chapter, we discuss more sophisticated options for encoding categorical predictors that address these issues. These options are available as recipe steps in the [`r pkg(embed)`](https://embed.tidymodels.org/) package.
+
+## Is an encoding necessary?
+
+A minority of models, such those based on trees or rules, can handle categorical data natively and do not require encoding or transformation of these kinds of features. A tree-based model can natively partition a variable like `Bldg_Type` into groups of factor levels, perhaps `OneFam` alone in one group and `Duplex` and `Twnhs` together in another group. Naive Bayes models are another example where the structure of the model can deal with categorical natively; distributions are computed within each level, for example, for all the different kinds of `Bldg_Type` in the data set.
+
+These models that can handle categorical features natively can _also_ deal with numeric, continuous features, making the transformation or encoding of such variables optional. Does this help? See Section 5.7 of @fes for a detailed exploration of how such models perform on benchmark data sets with untransformed factor variables compared with transformed dummy variables for those same features. In short, using dummy encodings did not typically result in better model performance but often required more time to train the models. 
+
+:::rmdnote
+We advise starting with untransformed categorical variables when a model allows it, and point out that more complex encodings often do not result in better performance for such models.
+:::
+
+## Using the outcome for encoding predictors
+
+There are multiple options for encodings more complex than dummy or indicator variables. One method called _effect_ or _likelihood encodings_ replaces the original categorical variables with a single numeric column that measures the effect of those data. For example, for the neighborhood predictor in the Ames housing data, we can compute the mean or median sale price for each neighborhood and substitute these means for the original data values. 
+
+```{r}
```

Comment:
I like this a lot more:

```r
ames_train %>%
  group_by(Neighborhood) %>%
  summarize(mean = mean(Sale_Price),
            std_err = sd(Sale_Price) / sqrt(length(Sale_Price))) %>% 
  ggplot(aes(y = reorder(Neighborhood, mean), x = mean)) + 
  geom_point() +
  geom_errorbar(aes(xmin = mean - 1.64 * std_err, xmax = mean + 1.64 * std_err)) +
  labs(y = NULL, x = "Price (mean, log scale)")
```

![image](https://user-images.githubusercontent.com/5731043/150562858-6c0d05d8-40c8-4be2-b348-30b74d92a8b7.png)